### PR TITLE
[Improvement][seatunnel-server] Improve the scheduler module to support more scheduling plugins

### PR DESCRIPTION
**Pull request by zhuangchong**
## Purpose of this pull request

Improve the scheduler module to support more scheduling plugins

1. From the perspective of the code module structure, SeaTunnel is to support scheduler plug-in, and can support more scheduling plug-ins, such as dolphinscheduler, azkaban

2. The current code does not really implement scheduler plug-in

Configuration file:

The key of the current property is ds.* . If azkaban scheduling is added, what should this key be?

https://github.com/apache/incubator-seatunnel/pull/2602/files#diff-be5fe744bda2308d461ce12ee32f8cc115dac0682230814f2b96e6ee0f5ade27L29-R43

Plugin implementation:

At present, spring dependency injection is used. When adding a plug-in module, an interface has two implementation classes, and the project will report an error at this time.

I introduced the ConditionalOnProperty annotation. According to the value of the configuration file seatunnel.scheduler.type, which implementation class should be injected into spring, only one implementation class will take effect in the project.

https://github.com/apache/incubator-seatunnel/pull/2602/files#diff-b0fc519f508b3019a859edab06e94430b51bd2df0b7c8cd834076d97f07155efR32-R33

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -206,6 +206,11 @@
                 <artifactId>seatunnel-config-shade</artifactId>
                 <version>${seatunnel.config.shade.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.apache.seatunnel</groupId>
+                <artifactId>seatunnel-scheduler-api</artifactId>
+                <version>${project.version}</version>
+            </dependency>
```
Comment (hailin0) [inline pom.xml]: Why declare submodules at root?

Comment (zhuangchong) [reply pom.xml]: ok, It has been modified.

**Comment by CalvinKirs** [comment]
Can you describe specifically what this PR does?

**Comment by zhuangchong** [comment]
> Can you describe specifically what this PR does?

Improve the scheduler module to support more scheduling plugins

1. From the perspective of the code module structure, SeaTunnel is to support scheduler plug-in, and can support more scheduling plug-ins, such as dolphinscheduler, azkaban

2. The current code does not really implement scheduler plug-in


Configuration file:

The key of the current property is ds.* . If azkaban scheduling is added, what should this key be?

https://github.com/apache/incubator-seatunnel/pull/2602/files#diff-be5fe744bda2308d461ce12ee32f8cc115dac0682230814f2b96e6ee0f5ade27L29-R43


Plugin implementation:

At present, spring dependency injection is used. When adding a plug-in module, an interface has two implementation classes, and the project will report an error at this time.

I introduced the ConditionalOnProperty annotation. According to the value of the configuration file seatunnel.scheduler.type, which implementation class should be injected into spring, only one implementation class will take effect in the project.

https://github.com/apache/incubator-seatunnel/pull/2602/files#diff-b0fc519f508b3019a859edab06e94430b51bd2df0b7c8cd834076d97f07155efR32-R33

**Comment by dijiekstra** [comment]
Great PR, I will review it at my leisure

File: seatunnel-server/seatunnel-app/pom.xml [hunk 2]
```diff
@@ -148,7 +148,7 @@
 
         <dependency>
             <groupId>org.apache.seatunnel</groupId>
-            <artifactId>seatunnel-spi</artifactId>
+            <artifactId>seatunnel-scheduler-api</artifactId>
```
Comment (dijiekstra) [inline seatunnel-server/seatunnel-app/pom.xml:151]: No need rename this module.We will have more interface definitions in  future , not just scheduling related

**Review by dijiekstra — CHANGES_REQUESTED** [review]
Can you start normally after you modify it? When I check out your PR, I can't launch the app properly. Please solve this problem first

**Review by ashulin — COMMENTED** [review]
The module has been moved to a subproject, can this PR be closed?

**Comment by hailin0** [comment]
> The module has been moved to a subproject, can this PR be closed?

+1

**Pull request closed without merging**